File: plugin/src/main/java/site/gutschi/documentation/jdeps/JDepsRunner.java
package site.gutschi.documentation.jdeps;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.spi.ToolProvider;

@SuppressWarnings({"unused", "UnusedReturnValue"})
public class JDepsRunner {
    private final ToolProvider jdepsTool = ToolProvider.findFirst("jdeps").orElseThrow();
    private final List<File> files = new LinkedList<>();
    private String multiRelease = null;
    private boolean ignoreMissingDeps = true;
    private boolean verbose = true;

    public static JDepsRunner create() {
        return new JDepsRunner();
    }

    public JDepsRunner file(String... fileNames) {
        Arrays.stream(fileNames)
                .map(File::new)
                .peek(this::ensureTargetFileValid)
                .forEach(this.files::add);
        return this;
    }

    public JDepsRunner fatJar(String fileName, Function<File, Boolean> matcher) {
        file(fileName);
        try (FileInputStream fis = new FileInputStream(fileName)) {
            Path tempFolder = FatJarUnzipper.unzip(fis);
            File libFolder = tempFolder.resolve("BOOT-INF").resolve("lib").toFile();
            if (!libFolder.exists()) {
                throw JDepsException.noFatJar(fileName);
            }
            Optional.of(libFolder)
                    .map(File::listFiles)
                    .stream()
                    .flatMap(Arrays::stream)
                    .filter(matcher::apply)
                    .forEach(this.files::add);
            return this;
        } catch (IOException e) {
            throw JDepsException.couldNotReadFatJar(e);
        }
    }

    public JDepsRunner multiRelease(String multiRelease) {
        this.multiRelease = multiRelease;
        return this;
    }

    public JDepsRunner verbose(boolean verbose) {
        this.verbose = verbose;
        return this;
    }

    public JDepsRunner ignoreMissingDeps(boolean ignoreMissingDeps) {
        this.ignoreMissingDeps = ignoreMissingDeps;
        return this;
    }

    public String run() {
        List<String> args = new LinkedList<>();
        if (multiRelease != null) {
            args.add("--multi-release");
            args.add(multiRelease);
        }
        if (ignoreMissingDeps) {
            args.add("--ignore-missing-deps");
        }
        if (verbose) {
            args.add("-v");
        }
        files.stream().map(File::getAbsolutePath).forEach(args::add);
        try(StringWriter stringWriter = new StringWriter()) {
            jdepsTool.run(new PrintWriter(stringWriter), new PrintWriter(System.err), args.toArray(String[]::new));
            return stringWriter.toString();
        } catch (Exception e) {
            throw JDepsException.wrap(e);
        }
    }

    private void ensureTargetFileValid(File targetFile) {
        if (!targetFile.exists()) {
            throw JDepsException.targetFileDoesNotExist(targetFile);
        }
        if (!Files.isReadable(targetFile.toPath())) {
            throw JDepsException.targetFileNotReadable(targetFile);
        }
    }
}

File: java-dependency-analyzer/src/main/kotlin/site/gutschi/dependency/jdeps/JDepsRunner.kt
package site.gutschi.dependency.jdeps

import java.io.File
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter
import java.nio.file.Files
import java.util.*
import java.util.spi.ToolProvider
import java.util.stream.Collectors

class JDepsRunner {
    private val jdepsTool = ToolProvider
        .findFirst("jdeps")
        .orElseThrow { JDepsException.jdepsNotFound() }
    private val files: MutableList<File> = LinkedList()
    private var fatJarMatcher: (File) -> Boolean = { true }
    private var multiRelease: String? = null

    fun fatJarMatcher(fatJarMatcher: (File) -> Boolean): JDepsRunner {
        if (files.size > 0) {
            throw JDepsException.matcherAfterFile()
        }
        this.fatJarMatcher = fatJarMatcher
        return this
    }

    fun file(vararg files: File): JDepsRunner {
        Arrays.stream(files).forEach { e: File -> this.singleFile(e) }
        return this
    }

    private fun singleFile(file: File) {
        ensureTargetFileValid(file)
        files.add(file)
        if (file.isDirectory || !file.name.endsWith(".jar")) {
            return
        }
        try {
            val fatJarReader = FatJarReader(file)
            if (!fatJarReader.isFatJar) {
                return
            }
            fatJarReader.streamJars()
                .filter { t: File -> fatJarMatcher(t) }
                .forEach { e: File -> this.singleFile(e) }
        } catch (e: IOException) {
            throw JDepsException.cannotUnzip(e)
        }
    }

    fun multiRelease(multiRelease: String): JDepsRunner {
        this.multiRelease = multiRelease
        return this
    }

    fun run(): JDepsResult {
        val args = getArgs()
        try {
            StringWriter().use { stringWriter ->
                val result = jdepsTool.run(PrintWriter(stringWriter), PrintWriter(System.err), *args.toTypedArray())
                if (result != 0) {
                    throw JDepsException.nonZeroReturn(result, stringWriter.toString())
                }
                return stringWriter.toString().lines().stream()
                    .filter { i -> isOutputLine(i) }
                    .map { l -> JDepsResultLine.fromLine(l) }
                    .collect(Collectors.toList())
            }
        } catch (e: Exception) {
            throw JDepsException.wrap(e)
        }
    }

    private fun isOutputLine(line: String): Boolean {
        return line.startsWith(" ")
    }

    private fun getArgs(): List<String> {
        val args: MutableList<String> = LinkedList()
        if (multiRelease != null) {
            args.add("--multi-release")
            args.add(multiRelease!!)
        }
        args.add("-v")
        args.add("-q")
        files.stream()
            .map { obj: File -> obj.absolutePath }
            .forEach { e: String -> args.add(e) }
        return args
    }

    private fun ensureTargetFileValid(targetFile: File) {
        if (!targetFile.exists()) {
            throw JDepsException.targetFileDoesNotExist(targetFile)
        }
        if (!Files.isReadable(targetFile.toPath())) {
            throw JDepsException.targetFileNotReadable(targetFile)
        }
    }
}